### ISSUE TITLE
CUDA-GCC version mapping for CUDA 11.4 

### ISSUE THREAD

**Issue by cdeepali** [pytorch/pytorch#81039]
### 🐛 Describe the bug

I saw the following error when building with CUDA v11.4.4 and GCC v11.2-

```
"<myenv>/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 843, in _check_cuda_version
    raise RuntimeError(
RuntimeError: The current installed version of <myenv>/_build_env/bin/powerpc64le-conda-linux-gnu-c++ (11.2.0) is greater than the maximum required version by CUDA 11.4 (10.0.0). Please make sure to use an adequate version of <myenv>/_build_env/bin/powerpc64le-conda-linux-gnu-c++ (>=6.0.0, <=10.0.0).
```

Looks like CUDA-GCC versions mapping at https://github.com/pytorch/pytorch/blob/master/torch/utils/cpp_extension.py#L56 needs to be updated for CUDA 11.4.1 and above. 

CUDA v11.4.4 requires GCC11 as per https://docs.nvidia.com/cuda/archive/11.4.4/cuda-installation-guide-linux/index.html#system-requirements.  

### Versions

Python version: 3.10.4
CUDA runtime version: 11.4.152
GCC: 11.2

cc @malfet @zou3519 @ngimel

**Comment by malfet** [comment]
Removing 1.21.1 tag, as CUDA seems to be completely ignoring semver assumptions: [11.4.0](https://docs.nvidia.com/cuda/archive/11.4.0/cuda-installation-guide-linux/index.html#system-requirements), when it was released supported up to gcc-9.x, but in [11.4.1](https://docs.nvidia.com/cuda/archive/11.4.1/cuda-installation-guide-linux/index.html#system-requirements) this quietly changes to gcc-11

I.e. revision version changes should only contain bugfixes, but 11.4 seems to have violated that and did a major compiler upgrade.

**Comment by hmaarrfk** [comment]
I wanted to cross reference a few other important discussions for better clarity on the underlying issue:

Most of my information comes from https://github.com/NVIDIA/nccl/issues/494#issuecomment-881074063= where the conclusions seem to be: 

* NVidia intended to bump the compiler support
* NVidia doesn't always update their documentation on time.
* A bug in gcc prevented this from happening in 10.3
* The bug was fixed in 10.4

The patch that you need to apply is pretty small
https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=a1b3484a8e6c53c8084723e3f1738d402374198e

We've backported this patch on conda-forge for gcc 10.3 and I don't think we've had any related issues with CUDA + patched 10.3

**Comment by hmaarrfk** [comment]
xref: https://github.com/pytorch/pytorch/issues/81680
